// @sap-cloud-sdk/odata-v4/dist/request/odata-action-import-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataActionImportRequestConfig = void 0;
const internal_1 = require("@sap-cloud-sdk/odata-common/internal");
/**
 * Action import request configuration for an entity type.
 * @typeParam DeSerializersT - Type of the deserializer use on the request
 * @typeParam ParametersT - Type of the parameter to setup a request with
 */
class ODataActionImportRequestConfig extends internal_1.ODataRequestConfig {
    /**
     * Creates an instance of ODataActionImportRequestConfig.
     * @param defaultServicePath - Default path of the service.
     * @param actionImportName - The name of the action import.
     * @param parameters - Parameters of the action imports.
     * @param oDataUri - URI conversion functions.
     */
    constructor(defaultServicePath, actionImportName, parameters, oDataUri) {
        super('post', defaultServicePath);
        this.actionImportName = actionImportName;
        this.parameters = parameters;
        this.oDataUri = oDataUri;
        this.payload = this.buildHttpPayload(parameters);
    }
    resourcePath() {
        return this.actionImportName;
    }
    queryParameters() {
        return {};
    }
    buildHttpPayload(parameters) {
        const payload = Object.keys(parameters).reduce((all, key) => {
            const payloadElement = parameters[key];
            if (typeof payloadElement.value !== 'undefined') {
                all[payloadElement.originalName] = payloadElement.value;
            }
            return all;
        }, {});
        return payload;
    }
}
exports.ODataActionImportRequestConfig = ODataActionImportRequestConfig;
