// @sap-cloud-sdk/generator/dist/request-builder/imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestBuilderImportDeclarations = void 0;
const ts_morph_1 = require("ts-morph");
const util_1 = require("@sap-cloud-sdk/util");
const imports_1 = require("../imports");
/**
 * @internal
 */
function requestBuilderImportDeclarations(entity, oDataVersion) {
    return [
        ...(0, imports_1.externalImportDeclarations)(entity.keys),
        (0, imports_1.odataImportDeclaration)([
            ...requestBuilderImports(entity),
            'DeserializedType',
            'RequestBuilder',
            ...(0, imports_1.propertyTypeImportNames)(entity.keys)
        ], oDataVersion),
        entityImportDeclaration(entity),
        ...entityKeyImportDeclaration(entity.keys)
    ];
}
exports.requestBuilderImportDeclarations = requestBuilderImportDeclarations;
function requestBuilderImports(entity) {
    const imports = [
        'DefaultDeSerializers',
        'DeSerializers',
        'GetAllRequestBuilder',
        'GetByKeyRequestBuilder'
    ];
    if (entity.creatable) {
        imports.push('CreateRequestBuilder');
    }
    if (entity.updatable) {
        imports.push('UpdateRequestBuilder');
    }
    if (entity.deletable) {
        imports.push('DeleteRequestBuilder');
    }
    return imports;
}
function entityImportDeclaration(entity) {
    return {
        kind: ts_morph_1.StructureKind.ImportDeclaration,
        namedImports: [entity.className],
        moduleSpecifier: `./${entity.className}`
    };
}
function entityKeyImportDeclaration(properties) {
    return (0, util_1.unique)(properties
        .filter(property => property.isEnum)
        .map(property => property.jsType)).map(type => ({
        kind: ts_morph_1.StructureKind.ImportDeclaration,
        namedImports: [type],
        moduleSpecifier: `./${type}`
    }));
}
