// @sap-cloud-sdk/generator/dist/vdm-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VdmUnsupportedReason = exports.VdmReturnTypeCategory = void 0;
/**
 * @internal
 */
var VdmReturnTypeCategory;
(function (VdmReturnTypeCategory) {
    VdmReturnTypeCategory[VdmReturnTypeCategory["ENTITY"] = 0] = "ENTITY";
    VdmReturnTypeCategory[VdmReturnTypeCategory["COMPLEX_TYPE"] = 1] = "COMPLEX_TYPE";
    VdmReturnTypeCategory[VdmReturnTypeCategory["EDM_TYPE"] = 2] = "EDM_TYPE";
    VdmReturnTypeCategory[VdmReturnTypeCategory["VOID"] = 3] = "VOID";
    VdmReturnTypeCategory[VdmReturnTypeCategory["NEVER"] = 4] = "NEVER";
})(VdmReturnTypeCategory = exports.VdmReturnTypeCategory || (exports.VdmReturnTypeCategory = {}));
/**
 * @internal
 */
var VdmUnsupportedReason;
(function (VdmUnsupportedReason) {
    VdmUnsupportedReason[VdmUnsupportedReason["ENTITY_NOT_DESERIALIZABLE"] = 0] = "ENTITY_NOT_DESERIALIZABLE";
})(VdmUnsupportedReason = exports.VdmUnsupportedReason || (exports.VdmUnsupportedReason = {}));
