// @sap-cloud-sdk/generator/dist/sdk-metadata/code-sample-util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOperationParams = exports.getOperationWithMinParameters = exports.getOperationWithoutParameters = exports.sampleOperationImport = exports.getShortestNameEntity = exports.getODataEntity = void 0;
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
/**
 * @internal
 */
function getODataEntity(serviceName, vdmEntities) {
    return ((0, internal_1.getLevenshteinClosest)(serviceName, vdmEntities, x => x.className) ||
        getShortestNameEntity(vdmEntities));
}
exports.getODataEntity = getODataEntity;
/**
 * @internal
 */
function getShortestNameEntity(vdmEntities) {
    // If no closest entity found, return the entity with shortest name
    return vdmEntities.sort((a, b) => a.className.length < b.className.length ? -1 : 1)[0];
}
exports.getShortestNameEntity = getShortestNameEntity;
/**
 * @internal
 */
function sampleOperationImport(serviceName, operationImports) {
    if (operationImports.length === 1) {
        return operationImports[0];
    }
    return ((0, internal_1.getLevenshteinClosest)(serviceName, operationImports, x => x.name) ||
        getOperationWithoutParameters(operationImports) ||
        getOperationWithMinParameters(operationImports));
}
exports.sampleOperationImport = sampleOperationImport;
/**
 * @internal
 */
function getOperationWithoutParameters(operationImports) {
    return operationImports.find(func => { var _a; return ((_a = func.parameters) === null || _a === void 0 ? void 0 : _a.length) === 0; });
}
exports.getOperationWithoutParameters = getOperationWithoutParameters;
/**
 * Sorts and gets a function or action import having minimum input parameters.
 * @param operationImports - function or action imports array
 * @returns Function or action containing minimum input parameters
 * @internal
 */
function getOperationWithMinParameters(operationImports) {
    const getOperations = operationImports.filter(func => { var _a; return ((_a = func.httpMethod) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === 'get'; });
    if (getOperations.length) {
        operationImports = getOperations;
    }
    const sortedOperations = operationImports.sort((funcA, funcB) => { var _a, _b; return ((_a = funcA.parameters) === null || _a === void 0 ? void 0 : _a.length) < ((_b = funcB.parameters) === null || _b === void 0 ? void 0 : _b.length) ? -1 : 1; });
    return sortedOperations[0];
}
exports.getOperationWithMinParameters = getOperationWithMinParameters;
/**
 * @internal
 */
function getOperationParams(parameters) {
    const paramString = parameters
        .slice(0, 2)
        .reduce((cumulator, currentParam) => `${cumulator}, ${currentParam.parameterName}: '${currentParam.parameterName}'`, '');
    return `{${paramString.substring(1)}${parameters.length > 2 ? ', ...' : ''} }`;
}
exports.getOperationParams = getOperationParams;
