// @sap-cloud-sdk/odata-common/dist/request/odata-batch-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataBatchRequestConfig = void 0;
const uuid_1 = require("uuid");
const odata_request_config_1 = require("./odata-request-config");
/**
 * OData batch request configuration for an entity type.
 */
class ODataBatchRequestConfig extends odata_request_config_1.ODataRequestConfig {
    /**
     * Creates an instance of ODataBatchRequestConfig.
     * @param defaultServicePath - The default OData service path.
     * @param boundary - Request boundary for separation of subrequests. Defaults to an autogenerated value.
     */
    constructor(defaultServicePath, boundary = `batch_${(0, uuid_1.v4)()}`) {
        super('post', defaultServicePath, {
            'content-type': `multipart/mixed; boundary=${boundary}`,
            accept: 'multipart/mixed'
        });
        this.defaultServicePath = defaultServicePath;
        this.boundary = boundary;
        this.subRequestPathType = 'relativeToService';
    }
    withSubRequestPathType(subRequestPathType) {
        this.subRequestPathType = subRequestPathType;
    }
    resourcePath() {
        return '$batch';
    }
    queryParameters() {
        return {};
    }
}
exports.ODataBatchRequestConfig = ODataBatchRequestConfig;
